sites/main-site/src/content/blog/2025/state-of-nf-core-CI.md

```diff
@@ -0,0 +1,86 @@
+
+So we started to search for yet another solution. This time both [Edmund](https://github.com/edmundmiller) and me stumbled upon [runsOn](https://runs-on.com/).
+This service provides an AWS CloudFormation template to create a self-hosted runner in your own AWS account. So it is basically doing what we did before, but without the terraform setup and comes with really good documentation.
+We switched switched the CI tests for [nf-core/modules](https://github.com/nf-core/modules/pull/7840) to this new solution a week before the March 2025 hackathon and it worked like a charm.
+It was so smooth I had to check several times that the tests are actually running on the self-hosted runners.
+
+Switching to the new solution was as simple as changing:
+
+```diff title="nf-core/modules/.github/workflows/nf-test.yml"
+- runs-on: "self-hosted"
++ runs-on:
++      - runs-on=${{ github.run_id }}
++      - runner=4cpu-linux-x64
+```
+
+And if we wanted to add GPU-enabled runners, we just specify a different runner in the workflow.
+Because we were already at it and because of the great effort by other nf-core members on adding `linux/arm64` containers to nf-core/modules, we also could quickly add ARM-based CI runners to the [nf-core/rnaseq pipeline tests](https://github.com/nf-core/rnaseq/pull/1530).
+
+An additional benefit compared to the previous terraform setup is that we now use spot instances, which are both cheaper and faster to start up.
+
+There were still some issues where the runners behaved differently compared to the GitHub-hosted runners.
+
+- We had problems accessing some public AWS resources (setting `aws.client.anonymous = true{:groovy}` in the `nextflow.config` fixed this)
+- Dependencies were missing for `setup-apptainer` (which was fixed by the [runsOn developer](https://github.com/runs-on/runs-on/releases/tag/v2.7.0)).
+
+But overall this new runner setup works really well.
+
+## 2. nf-test sharding
+
+A further approach on speeding up the CI tests was introduced in [nf-test 0.9.0](https://github.com/askimed/nf-test/releases/tag/v0.9.0).
+This release added the option to split up the tests into multiple jobs using the `--shard` flag.
+Instead of running all tests one after the other, nf-test will now distribute the tests into multiple jobs and run them in parallel based on given maximum number of shards.
+This is especially useful for modules/subworkflows/pipelines that trigger a large number of tests.
+
+Let's look at the tests for the [FASTQC module](https://github.com/nf-core/modules/blob/master/modules/nf-core/fastqc/tests/main.nf.test) as an example. This module contains 12 tests, e.g. `sarscov2 single-end [fastq]`,`sarscov2 paired-end [fastq]` and their stub versions (and it is also included in 2 subworkflows, which would get tested during an update).
+With sharding enabled, we can now run, for example, 4 shards/jobs in parallel, each with 3 tests and the tests from the subworkflows distributed over them.
```

Comment:
```suggestion
With sharding enabled, we can now run, for example, 4 shards/jobs in parallel, each with 3 tests and the tests from the subworkflows distributed over them.
This greatly reduces the number of separate GitHub Actions runners that we need to launch,
which helps with concurrency limits and the resulting queue time.
```

Reply:
sharding doesn't tackle so much the runner problem (because before we just needed 1 runner vs. max_shards runners now. I will rephrase this.